src/python/pants/backend/python/subsystems/python_native_code.py

```diff
@@ -129,3 +139,157 @@ def check_build_for_current_platform_only(self, targets):
       'native code. Please ensure that the platform arguments in all relevant targets and build '
       'options are compatible with the current platform. Found targets for platforms: {}'
       .format(str(platforms_with_sources)))
+
+
+@rule(CToolchain, [Select(PythonNativeCode)])
+def select_c_toolchain_for_local_dist_compilation(python_native_code):
+  llvm_c_toolchain = yield Get(LLVMCToolchain, NativeToolchain, python_native_code.native_toolchain)
+  yield llvm_c_toolchain.c_toolchain
+
+
+@rule(CppToolchain, [Select(PythonNativeCode)])
+def select_cpp_toolchain_for_local_dist_compilation(python_native_code):
+  llvm_cpp_toolchain = yield Get(
+    LLVMCppToolchain, NativeToolchain, python_native_code.native_toolchain)
+  yield llvm_cpp_toolchain.cpp_toolchain
+
+
+class SetupPyNativeTools(datatype([
+    ('c_toolchain', CToolchain),
+    ('cpp_toolchain', CppToolchain),
+    ('platform', Platform),
+])):
+  """The native tools needed for a setup.py invocation.
+
+  This class exists because `SetupPyExecutionEnvironment` is created manually, one per target.
+  """
+
+
+# TODO: could this kind of @rule be automatically generated?
```

Comment:
Any time you have a rule like this I think its a sign you don't need a rule like this - for 2 reasons:
1. You're likely creating a middleman wrapper struct - if so, let the folks needing the struct components ask for them directly.
2. You're just running non-io python code - this is generally an abuse of parallelism and in-efficient, just run that code where you need to run it, not off in a rule.

Reply:
This was a very helpful comment! Along these lines, I was actually able to remove all the rules from the python backend in 9846475 and request `LLVMCToolchain`/etc directly.